### PR TITLE
feat: Use file cache query builder for searching for favorites

### PR DESCRIPTION
**Pull request by juliusknorr**
When trying to REPORT files that are favorites we still used the old code path to aggregate them, this moves this over to the filecache search query builder methods.

Request to trigger:

```
blackfire curl -X REPORT \
  --url http://admin:admin@nextcloud.local/remote.php/dav/files/admin/MyGroupfolder4 \
  --header 'Content-Type: application/xml' \
  --data '<?xml version="1.0"?>
<oc:filter-files  xmlns:d="DAV:" xmlns:oc="http://owncloud.org/ns" xmlns:nc="http://nextcloud.org/ns">
         <oc:filter-rules>
                 <oc:favorite>1</oc:favorite>
         </oc:filter-rules>
        <d:prop>
            <d:resourcetype />
        </d:prop>
 </oc:filter-files> '
```

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/FilesReportPlugin.php [hunk 1]
```diff
@@ -363,6 +364,33 @@ protected function processFilterRulesForFileNodes(array $filterRules, ?int $limi
 			}
 		}
 
+		$ns = '{' . $this::NS_OWNCLOUD . '}';
```
Comment (blizzz) [inline apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:367]: add as const like it is for systemtags and circles.

File: apps/dav/lib/Connector/Sabre/FilesReportPlugin.php [hunk 2]
```diff
@@ -363,6 +364,33 @@ protected function processFilterRulesForFileNodes(array $filterRules, ?int $limi
...
+		$favoriteFilter = null;
+		foreach ($filterRules as $filterRule) {
+			if ($filterRule['name'] === $ns . 'favorite') {
+				$favoriteFilter = true;
```
Comment (blizzz) [inline apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:371]: can break the loop afterwards (will not gain significant speed, of course, but still no reason to waste cycles as principle)

File: apps/dav/lib/Connector/Sabre/FilesReportPlugin.php [hunk 3]
```diff
@@ -363,6 +364,33 @@ protected function processFilterRulesForFileNodes(array $filterRules, ?int $limi
...
+			}
+			if ($nodes === []) {
+				// there cannot be a common match when nodes are empty early.
+				return $nodes;
+			}
+
+			if ($limit !== null || $offset !== null) {
+				$nodes = array_slice($nodes, $offset, $limit);
+			}
+		}
```
Comment (blizzz) [inline apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:392]: Duplicated code, move it to a separate func

File: lib/private/Files/Node/Folder.php [hunk 4]
```diff
@@ -295,8 +295,8 @@ public function searchByMime($mimetype) {
 	 * @param string $userId owner of the tags
 	 * @return Node[]
 	 */
-	public function searchByTag($tag, $userId) {
-		$query = $this->queryFromOperator(new SearchComparison(ISearchComparison::COMPARE_EQUAL, 'tagname', $tag), $userId);
+	public function searchByTag($tag, $userId, int $limit = 0, int $offset = 0) {
```
Comment (blizzz) [inline lib/private/Files/Node/Folder.php:298]: I suppose the interface needs to be extended as well? Though I don't see a psalm warning.
LazyFolder should look the same, alone for consistency.

**Review by blizzz — COMMENTED** [review]
The favorites logic in `processFilterRulesForFileIDs()` should be removed, also. 

All the comments are no-brainers and I understand due to the fact this being an early draft. The approach is fine!

File: lib/private/Files/Node/Folder.php (hunk 4 above)
Comment (juliushaertl) [reply lib/private/Files/Node/Folder.php:298]: Adjusted, would need another adjustment in groupfolders again once merged.

For potential backports this is a bit more tricky but we may just use an ugly direct call on the private class implementation of a separate method then to avoid the interface adjustment.

**Comment by juliushaertl** [comment]
Ready for review now

File: lib/private/Files/Node/Folder.php (hunk 4 above)
Comment (blizzz) [reply lib/private/Files/Node/Folder.php:298]: The foundational PRs were also backported and there the `Folder` changes were indeed made private. Since the signature is different, it's likely taking more to make at least tooling accept that state.

File: lib/private/Files/Node/NonExistingFolder.php [hunk 5]
```diff
@@ -150,7 +150,7 @@ public function searchByMime($mimetype) {
 		throw new NotFoundException();
 	}
 
-	public function searchByTag($tag, $userId) {
+	public function searchByTag($tag, $userId, int $limit = 0, int $offset = 0) {
```
Comment (blizzz) [inline lib/private/Files/Node/NonExistingFolder.php:153]: That class is a good finding, i shall add searchBySystemTag here also – will be a separate PR.

**Review by blizzz — APPROVED** [review]

File: apps/dav/lib/Connector/Sabre/FilesReportPlugin.php [hunk 6]
```diff
@@ -345,27 +330,57 @@ protected function processFilterRulesForFileNodes(array $filterRules, ?int $limi
...
 			}
 
 			if (!$oneTagSearch && ($limit !== null || $offset !== null)) {
-				$nodes = array_slice($nodes, $offset, $limit);
+				$nodes = array_slice($nodes, $offset ?? 0, $limit);
+			}
+		}
+
+		if ($this->hasFilterFavorites($filterRules)) {
+			$tmpNodes = $this->userFolder->searchByTag(ITags::TAG_FAVORITE, $this->userSession->getUser()->getUID(), $limit ?? 0, $offset ?? 0);
```
Comment (marcelklehr) [inline apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: Isn't this this thing again where we slice multiple times and then intersect the results? Does this give proper results? I forgot.

Comment (marcelklehr) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]:
```
limit = 3
offset = 2

a = [0,3,6,9,12,15,18,21,24]
b = [0,2,4,6,8,10,12,14,16,18,20,22,24]

slice(a) // 6,9,12
slice(b) // 4,6,8

intersect(slice(a), slice(b)) // 6

intersect(a,b) // 0,6,12,18,24
slice(intersect(a,b)) // 12,18,24
```

Comment (marcelklehr) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: So, it can give too little results. Sometimes even none when some would be expected.

Comment (juliushaertl) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: Good point, will see what could be done about that

Comment (icewind1991) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: Since both the `searchBySystemTag` and `searchByTag` end up doing a `search` with their specific search queries anyway. You can instead build a single search query that performs all the filtering in one go.

Comment (icewind1991) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: and you can probably also include the filtering for the fileids from `processFilterRulesForFileIDs` in the same query to remove the need for the `findNodesByFileIds` later down

Comment (marcelklehr) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: > You can instead build a single search query that performs all the filtering in one go.

Search stuff is not in OCP, though. Would be awesome if it was.

Comment (icewind1991) [reply apps/dav/lib/Connector/Sabre/FilesReportPlugin.php:346]: using it in the dav app is fine since it's releases are coupled with core
